core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -74,255 +61,286 @@
+         Completable storeStartup = Flowable.fromIterable(configuration.persistence().stores())
+               // We have to ensure stores are started in configured order to ensure the stores map retains that order
+               .concatMapSingle(storeConfiguration -> {
+                  NonBlockingStore<?, ?> actualStore = storeFromConfiguration(storeConfiguration);
+                  NonBlockingStore<?, ?> nonBlockingStore;
+                  if (storeConfiguration.async().enabled()) {
+                     nonBlockingStore = new AsyncNonBlockingStore<>(actualStore);
+                  } else {
+                     nonBlockingStore = actualStore;
+                  }
+                  StoreConfiguration processedConfiguration = cacheStoreFactoryRegistry.processStoreConfiguration(storeConfiguration);
+                  InitializationContextImpl ctx =
+                        new InitializationContextImpl(processedConfiguration, cache.wired(), keyPartitioner, persistenceMarshaller,
+                              timeService, byteBufferFactory, marshallableEntryFactory, nonBlockingExecutor, globalConfiguration, blockingManager);
+                  CompletionStage<Void> stage = nonBlockingStore.start(ctx);
+                  return Completable.fromCompletionStage(stage)
+                        .toSingle(() -> new StoreStatus(nonBlockingStore, processedConfiguration,
+                              updateCharacteristics(nonBlockingStore, nonBlockingStore.characteristics(), storeConfiguration)));
+               })
+               // This relies upon visibility guarnatees of reactive streams for publishing map values
+               .doOnNext(status -> stores.put(status.store, status))
+               .delay(status -> {
+                  if (status.config.purgeOnStartup()) {
+                     return Flowable.fromCompletable(Completable.fromCompletionStage(status.store.clear()));
+                  }
+                  return Flowable.empty();
+               })
+               .ignoreElements();
+
+         long interval = configuration.persistence().availabilityInterval();
+         if (interval > 0) {
+            storeStartup = storeStartup.doOnComplete(() ->
+               availabilityTask = nonBlockingManager.scheduleWithFixedDelay(this::pollStoreAvailability, interval, interval, MILLISECONDS));
          }
-      } catch (Exception e) {
-         throw new CacheException("Unable to start cache loaders", e);
-      }
-   }
 
-   /**
-    * Returns how many publisher invocations are currently active.
-    * @return count of active publisher instances
-    */
-   public int activePublisherInvocations() {
-      return Integer.MAX_VALUE - publisherSemaphore.availablePermits();
+         storeStartup.doOnComplete(() -> lock.unlockWrite(stamp))
+               // Blocks here waiting for stores and availability task to start if needed
+               .blockingAwait();
+      } catch (Throwable t) {
+         lock.unlockWrite(stamp);
```

Comment:
Should we add a trace log of the exception before propogating?

Reply:
Sure, I don't see why not :)